File: deconvolution/matlab/bold_deconvolution.m
function neuro = bold_deconvolution(BOLD,TR,alpha,NT,par,xb,Hxb)

% Deconvolves a preprocessed BOLD signal into neuronal time series 
% based on discrete cosine set and ridge regression.
% This function does not use confound regressors (e.g. motion).
% This function does not perform whitening and temporal filtering.
% The BOLD input signal must already be pre-processed.
%
% FORMAT neuro = ridge_regress_deconv(BOLD,TR)
%
% BOLD  - Preprocessed BOLD signal (time points X ROIs)
% TR    - Time repetition, [s]
%
% FORMAT neuro = bold_deconvolution(BOLD,TR,alpha,NT,par,xb,Hxb)
%
% Optional inputs:
%
% alpha - Regularization parameter 
%         (default: 0.005)
% NT    - Microtime resolution (number of time bins per scan)
%         (default: 16)
% par   - Parallel or sequential computations
%         (default: 0)
% xb    - Temporal basis set in microtime resolution 
%         (default: discrete cosine set)
% Hxb   - Convolved temporal basis set in scan resolution
%         (default: discrete cosine set convolved with canonical HRF)
% _________________________________________________________________________
% Copyright (C) 2024 Ruslan Masharipov
% Contact email: masharipov@ihb.spb.ru

% Setup variables
if nargin < 2
    error('Define time repetition (TR)')
elseif nargin < 3
    alpha = 0.005;
    NT = 16;
    par = 0;
elseif nargin < 4
    NT = 16;
    par = 0;
elseif nargin < 5
    par = 0;
elseif nargin == 6
    error('Define convolved temporal basis set (Hxb)')
end

dt = TR/NT;                      % Length of time bin, [s]
N = size(BOLD,1);                % Scan duration, [dynamics] 
k = 1:NT:N*NT;                   % Microtime to scan time indices

if exist('Hxb', 'var') ~= 1

    % Create canonical HRF in microtime resolution (identical to SPM cHRF)
    t = 0:dt:32;
    hrf = gampdf(t,6) - gampdf(t,NT)/6;
    hrf = hrf'/sum(hrf);
    
    % Create convolved discrete cosine set
    M = N*NT + 128;
    n = (0:(M -1))';
    xb = zeros(size(n,1),N);
    xb(:,1) = ones(size(n,1),1)/sqrt(M);
    for j=2:N
        xb(:,j) = sqrt(2/M)*cos(pi*(2*n+1)*(j-1)/(2*M));
    end
    
    Hxb = zeros(N,N);
    for i = 1:N
        Hx       = conv(xb(:,i),hrf);
        Hxb(:,i) = Hx(k + 128);
    end
    xb = xb(129:end,:);

end

switch par
    case 0
        for ROI = 1:size(BOLD,2)
            % Perform ridge regression
            C(:,ROI) = (Hxb'*Hxb + alpha*eye(length(Hxb)))\(Hxb'*BOLD(:,ROI));
            
            % Recover neuronal signal
            neuro(:,ROI) = xb*C(:,ROI);
        end
    case 1
        parfor ROI = 1:size(BOLD,2)
            % Perform ridge regression
            C(:,ROI) = (Hxb'*Hxb + alpha*eye(length(Hxb)))\(Hxb'*BOLD(:,ROI));
            
            % Recover neuronal signal
            neuro(:,ROI) = xb*C(:,ROI);
        end
end

File: deconvolution/python/bold_deconvolution.py
import numpy as np
from scipy.stats import gamma

def ridge_regress_deconvolution(BOLD: np.ndarray,
                                TR: float,
                                alpha: float = 0.005,
                                NT: int = 16,
                                xb: np.ndarray = None,
                                Hxb: np.ndarray = None) -> np.ndarray:
    """
    Deconvolves a preprocessed BOLD signal into neuronal time series
    based on selected temporal basis set (by default: discrete cosine
    set convolved with canonical HRF) and ridge regression.

    This function does not use confound regressors (e.g., motion).
    This function does not perform whitening and temporal filtering.
    The BOLD input signal must already be pre-processed.

    :param BOLD: Preprocessed BOLD signal (numpy array)
    :type BOLD: numpy.ndarray (NxROI)
    :param TR: Time repetition, in seconds
    :type TR: float
    :param alpha: Regularization parameter, defaults to 0.005
    :type alpha: float, optional
    :param NT: Microtime resolution (number of time bins per scan), defaults to 16
    :type NT: int, optional
    :param xb: Temporal basis set in microtime resolution
    :type xb: numpy.ndarray (N*NTxN*NT)
    :param Hxb: Convolved temporal basis set in scan resolution
    :type Hxb: numpy.ndarray (NxN)
    :return: Deconvolved neuronal time series
    :rtype: numpy.ndarray

    :raises ValueError: If the BOLD signal is empty
    :raises ZeroDivisionError: If the time repetition (TR) is zero
    """

    if len(BOLD) == 0:
        raise ValueError('BOLD signal is empty.')
    if TR == 0:
        raise ZeroDivisionError('TR should be more than 0')
    dt = TR / NT   # Length of time bin, [s]
    N = len(BOLD)  # Scan duration, [dynamics]
    k = np.arange(0, N * NT, NT)  # Microtime to scan time indices

    if xb is None and Hxb is not None:
        raise ValueError("Both xb and Hxb should be specified, or neither of them")
    if Hxb is None and xb is not None:
        raise ValueError("Both xb and Hxb should be specified, or neither of them")

    if xb is None:
        xb, Hxb = compute_xb_Hxb(N, NT, TR) # Use discrete cosine set convolved with canonical HRF

    # Perform ridge regression
    C = np.linalg.solve(Hxb.T @ Hxb + alpha * np.eye(N), Hxb.T @ BOLD)

    # Recover neuronal signal
    neuro = xb @ C

    return neuro.flatten()


def compute_xb_Hxb(N: int, NT: int, TR: float) -> np.ndarray:
    dt = TR / NT
    k = np.arange(0, N * NT, NT)  # Microtime to scan time indices

    # Create canonical HRF in microtime resolution (identical to SPM12 cHRF)
    t = np.arange(0, 32 + dt, dt)
    hrf = gamma.pdf(t, 6) - gamma.pdf(t, NT) / 6
    hrf = hrf / np.sum(hrf)

    # Create convolved discrete cosine set
    M = N * NT + 128
    xb = dctmtx_numpy_vect(M, N)
    Hxb = np.zeros((N, N))
    for i in range(N):
        Hx = np.convolve(xb[:, i], hrf, mode='full')
        Hxb[:, i] = Hx[k + 128]
    xb = xb[128:, :]
    return xb, Hxb


def dctmtx_numpy_vect(N: int, K: int) -> np.ndarray:
    n = np.arange(N)
    C = np.zeros((N, K))
    C[:, 0] = 1 / np.sqrt(N)
    k = np.arange(1, K)
    C[:, 1:K] = np.sqrt(2 / N) * np.cos(np.pi * (2 * n[:, np.newaxis]) * k / (2 * N))
    return C


